src/libc/locale/FreeBSD/wctype.3
.\" Copyright (c) 2002 Tim J. Robbins
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: src/lib/libc/locale/wctype.3,v 1.7 2006/10/13 16:11:12 ru Exp $
.\"
.Dd March 27, 2004
.Dt WCTYPE 3
.Os
.Sh NAME
.Nm iswctype ,
.Nm iswctype_l ,
.Nm wctype ,
.Nm wctype_l
.Nd "wide character class functions"
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In wctype.h
.Ft int
.Fo iswctype
.Fa "wint_t wc"
.Fa "wctype_t charclass"
.Fc
.Ft wctype_t
.Fo wctype
.Fa "const char *property"
.Fc
.In wctype.h
.In xlocale.h
.Ft int
.Fo iswctype_l
.Fa "wint_t wc"
.Fa "wctype_t charclass"
.Fa "locale_t loc"
.Fc
.Ft wctype_t
.Fo wctype_l
.Fa "const char *property"
.Fa "locale_t loc"
.Fc
.Sh DESCRIPTION
The
.Fn wctype
function returns a value of type
.Vt wctype_t ,
which represents the requested wide character class and
may be used as the second argument for calls to
.Fn iswctype .
.Pp
The following character class names are recognised:
.Bl -column -offset indent ".Li alnum" ".Li cntrl" ".Li ideogram" ".Li print" ".Li space"
.It Li "alnum	cntrl	ideogram	print	space	xdigit"
.It Li "alpha	digit	lower	punct	special"
.It Li "blank	graph	phonogram	rune	upper"
.El
.Pp
The
.Fn iswctype
function checks whether the wide character
.Fa wc
is in the character class
.Fa charclass .
.Pp
Although the
.Fn iswctype
and
.Fn wctype
functions use the current locale, the
.Fn iswctype_l
and
.Fn wctype_l
functions may be passed locales directly. See
.Xr xlocale 3
for more information.
.Sh RETURN VALUES
The
.Fn iswctype
function returns non-zero if and only if
.Fa wc
has the property described by
.Fa charclass ,
or
.Fa charclass
is zero.
.Pp
The
.Fn wctype
function returns 0 if
.Fa property
is invalid; otherwise, it returns a value of type
.Vt wctype_t
that can be used in subsequent calls to
.Fn iswctype .
.Sh EXAMPLES
Reimplement
.Xr iswalpha 3
in terms of
.Fn iswctype
and
.Fn wctype :
.Bd -literal -offset indent
int
myiswalpha(wint_t wc)
{
	return (iswctype(wc, wctype("alpha")));
}
.Ed
.Sh SEE ALSO
.Xr ctype 3 ,
.Xr nextwctype 3 ,
.Xr xlocale 3
.Sh STANDARDS
The
.Fn iswctype
and
.Fn wctype
functions conform to
.St -p1003.1-2001 .
The
.Dq Li ideogram ,
.Dq Li phonogram ,
.Dq Li special ,
and
.Dq Li rune
character classes are extensions.
.Sh HISTORY
The
.Fn iswctype
and
.Fn wctype
functions first appeared in
.Fx 5.0 .


libc/locale/FreeBSD/wctype.c
/*-
 * Copyright (c) 2002 Tim J. Robbins.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libc/locale/wctype.c,v 1.4 2008/03/17 18:22:23 antoine Exp $");

#include "xlocale_private.h"

#include <ctype.h>
#include <string.h>
#include <wctype.h>
#include <limits.h>

wctype_t
wctype_l(const char *property, locale_t loc)
{
	_RuneLocale *rl;
	static const struct {
		const char	*name;
		wctype_t	 mask;
	} props[] = {
		{ "alnum",	_CTYPE_A|_CTYPE_D },
		{ "alpha",	_CTYPE_A },
		{ "blank",	_CTYPE_B },
		{ "cntrl",	_CTYPE_C },
		{ "digit",	_CTYPE_D },
		{ "graph",	_CTYPE_G },
		{ "lower",	_CTYPE_L },
		{ "print",	_CTYPE_R },
		{ "punct",	_CTYPE_P },
		{ "space",	_CTYPE_S },
		{ "upper",	_CTYPE_U },
		{ "xdigit",	_CTYPE_X },
		{ "ideogram",	_CTYPE_I },	/* BSD extension */
		{ "special",	_CTYPE_T },	/* BSD extension */
		{ "phonogram",	_CTYPE_Q },	/* BSD extension */
		{ "rune",	0xFFFFFF00L },	/* BSD extension */
		{ NULL,		0UL },		/* Default */
	};
	int i;

	i = 0;
	while (props[i].name != NULL && strcmp(props[i].name, property) != 0)
		i++;

	if (props[i].mask)
		return (props[i].mask);

	NORMALIZE_LOCALE(loc);
	rl = &loc->__lc_ctype->_CurrentRuneLocale;
	if ((i = rl->__ncharclasses) > 0) {
		_RuneCharClass *rp;
		for (rp = rl->__charclasses; i-- > 0; rp++) {
			if (strncmp(rp->__name, property, CHARCLASS_NAME_MAX) == 0)
				return (rp->__mask);
		}
	}
	return 0;
}

wctype_t
wctype(const char *property)
{
	return wctype_l(property, __current_locale());
}
